### ISSUE TITLE
UselessCallOnNotNull false positive with listOfNotNull

### ISSUE THREAD

**Issue by SebRut** [detekt/detekt#5077]
## Expected Behavior
Detekt should not report `UselessCallOnNotNull` on `listOfNotNull` in the code snippet in "Steps to Reproduce"

## Observed Behavior
Detekt reports `UselessCallOnNotNull` on the `listOfNotNull` in the snippet.

## Steps to Reproduce
```kotlin
fun getAvailableMethods(): Array<Method> {
        return listOfNotNull(
                Method.M,
                Method.P.takeIf { service.isAvailable() }
        ).toTypedArray()
}
```

## Context
We use detekt in multi-module android project and apply it using a `detektAll` task. Config is built upon default.

## Your Environment

* Version of detekt used: 1.20.0
* Version of Gradle used (if applicable):7.3.1
* Operating System and version: macOS 11.6.7 (BigSur) but also in CI

**Comment by BraisGabin** [comment]
I can't create a unit test with your snippet because there are a lot of unknown classes but I implemented a similar version:

```kotlin
fun getAvailableMethods(): Array<String> {
    return listOfNotNull(
        "Method.M",
        "Method.P".takeIf { false }
    ).toTypedArray()
}
```

And it works as expected. No issues are raised.

Could you provide a reproducer?

**Comment by SebRut** [comment]
> Could you provide a reproducer?

I've created a repro repo under https://github.com/SebRut/detekt-issue-5077-repro . Running the `detektAll` gradle tasks leads to the issue

**Comment by 3flex** [comment]
Thanks for the reproducer, that helps.

The problem here is that in the `detektAll` task you've created you're also setting a value for classpath:
https://github.com/SebRut/detekt-issue-5077-repro/blob/bc36c0f56ca116fbddcb2fd96b5da737aa2494d4/build.gradle#L25

This enables rules that require type resolution, however this also requires the classpath to be set appropriately so all types can be resolved by detekt. In this case you're setting the classpath to the detekt configuration, which doesn't make sense - the classpath should be set according to the needs of your specific project, or instead, you can use the preconfigured tasks that detekt Gradle plugin creates. In your project you should apply plugin `id "io.gitlab.arturbosch.detekt"` to your `app` module then run `./gradlew :app:detektMain`. You'll see this, correctly, doesn't report an issue, because detekt auto-configured the correct classpath for that task, so the UselessCallOnNotNull rule works correctly.

I'll leave this open for now and will close in a few days if you don't report back - but please let me know if I've misunderstood anything about your issue or reproducer. In the meantime if you need some help with setting up detekt, please try Slack or opening a GitHub discussion.

**Comment by SebRut** [comment]
Hey, sorry for taking that long to respond.

The detekt configuration in the repro project is adopted from our internal, multi-module project. Applying detekt on the `app` module level leads to it only running on source from the module. I also tried applying detekt to the module containing the "false-positive" code from this issue which seems to fix the false-positive issue.

**Comment by 3flex** [comment]
Just reviewing this old issue.

Is there anything else you need help with? I don't think there's an issue here, though perhaps the documentation can be improved.

**Comment by cortinico** [comment]
Closing as there is no actionable here.
